core/src/main/java/org/opensearch/remote/metadata/common/SdkClientUtils.java

```diff
@@ -30,12 +31,22 @@ private SdkClientUtils() {}
     /**
      * Unwraps the cause of a {@link CompletionException}. If the cause is an {@link Exception}, rethrows the exception.
      * Otherwise wraps it in an {@link OpenSearchException}. Properly re-interrupts the thread on {@link InterruptedException}.
-     * @param throwable a throwable, expected to be a {@link CompletionException} or {@link CancellationException}.
+     * @param throwable a throwable.
+     * @param exceptionTypesToUnwrap optional list of exception types to unwrap. Defaults to {@link CompletionException}.
      * @return the cause of the completion exception or the throwable, directly if an {@link Exception} or wrapped in an OpenSearchException otherwise.
      */
-    public static Exception unwrapAndConvertToException(Throwable throwable) {
-        // Unwrap completion exception or pass through other exceptions
-        Throwable cause = throwable instanceof CompletionException ? throwable.getCause() : throwable;
+    @SafeVarargs
+    public static Exception unwrapAndConvertToException(Throwable throwable, Class<? extends Throwable>... exceptionTypesToUnwrap) {
+        // Unwrap specified exception types or pass through other exceptions
+        List<Class<? extends Throwable>> unwrapTypes = (exceptionTypesToUnwrap.length > 0)
```

Comment:
Sorry this could be a noob question, but remind me again What happens if the exception chain is deeper than one level?

Reply:
Also from ml-commos or flow framework, we need to call like:

`Exception ex = SdkClientUtils.unwrapAntdConvertToException(t, OpenSearchStatusException.class);` regardless of if we want to catch the error IndexNotFoundException or [VersionConflictEngineException](https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java#L158C79-L158C109)

Reply:
It will iterate the nested exceptions until it finds one that’s not in the list or until the cause is null. 

And yes for now that’s how you’d need to call it.  I will be adding some convenience methods to the abstract class that do that for you though.